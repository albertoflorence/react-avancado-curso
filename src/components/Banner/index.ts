export { default } from './Banner'
export * from './Banner'
export * as BannerStyles from './BannerStyles'
