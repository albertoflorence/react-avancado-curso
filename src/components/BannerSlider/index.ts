export { default } from './BannerSlider'
export * from './BannerSlider'
export * as BannerSliderStyles from './BannerSliderStyles'
