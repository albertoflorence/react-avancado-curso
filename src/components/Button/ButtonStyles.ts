import styled, { css, DefaultTheme } from 'styled-components'
import { ButtonProps } from './Button'

const sizes = {
  small: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.small};
  `,
  medium: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacing(1, 4)};
  `,
  large: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.large};
    padding: ${theme.spacing(1, 6)};
  `
}

const handleFullWidth = () => css`
  width: 100%;
`

const handleIcon = (theme: DefaultTheme) => css`
  display: inline-flex;
  align-items: center;
  justify-content: center;

  svg {
    width: 1em;
    & + span {
      margin-left: ${theme.spacing(1)};
    }
  }

  span + svg {
    margin-left: ${theme.spacing(1)};
  }
`

interface WrapperProps extends ButtonProps {
  hasIcon: boolean
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size = 'medium', fullWidth = false, hasIcon }) => css`
    background: ${theme.gradients.primary};
    color: ${theme.colors.white};
    border: none;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacing(1)};
    cursor: pointer;

    ${size && sizes[size](theme)}
    ${fullWidth && handleFullWidth}
    ${hasIcon && handleIcon(theme)}

    &:hover {
      filter: brightness(95%);
    }
  `}
`
