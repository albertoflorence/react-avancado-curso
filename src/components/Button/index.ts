export { default } from './Button'
export * from './Button'
export * as ButtonStyles from './ButtonStyles'
