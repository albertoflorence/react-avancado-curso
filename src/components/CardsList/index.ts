export { default } from './CardsList'
export * from './CardsList'
export * as CardsListStyles from './CardsListStyles'
