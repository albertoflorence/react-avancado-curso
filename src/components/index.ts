export { default as Icon } from './Icon/Icon'
export { default as Logo } from './Logo/Logo'
export { default as MediaMatch } from './MediaMatch/MediaMatch'
export { default as Heading } from './Heading/Heading'
export { default as Link } from './Link/Link'
