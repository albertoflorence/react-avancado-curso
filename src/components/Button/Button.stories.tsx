import { Story, Meta } from '@storybook/react'
import Button, { ButtonProps } from './Button'
import { AddShoppingCart } from 'components/Icons'
export default {
  title: 'Button',
  component: Button
} as Meta

export const Default: Story<ButtonProps> = args => <Button {...args} />
Default.args = {
  children: 'Buy now'
}

export const WithStartIcon: Story<ButtonProps> = args => <Button {...args} />
WithStartIcon.args = {
  children: 'Buy now',
  startIcon: <AddShoppingCart />
}

export const WithEndIcon: Story<ButtonProps> = args => <Button {...args} />
WithEndIcon.args = {
  children: 'Buy now',
  endIcon: <AddShoppingCart />
}
