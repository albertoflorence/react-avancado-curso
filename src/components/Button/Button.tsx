import * as S from './ButtonStyles'

export interface ButtonProps extends React.ComponentPropsWithoutRef<'button'> {
  children?: React.ReactNode
  size?: 'small' | 'medium' | 'large'
  fullWidth?: boolean
  startIcon?: React.ReactNode
  endIcon?: React.ReactNode
}

const Button = ({ children, startIcon, endIcon, ...props }: ButtonProps) => (
  <S.Wrapper {...props} hasIcon={Boolean(startIcon || endIcon)}>
    {startIcon}
    {children && <span> {children} </span>}
    {endIcon}
  </S.Wrapper>
)

export default Button
