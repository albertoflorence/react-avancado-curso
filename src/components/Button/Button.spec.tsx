import { screen } from '@testing-library/react'
import { renderWithTheme, theme } from 'utils/tests'
import Button, { ButtonProps } from './Button'
import { Icon } from 'components'

const makeIcon = (id: string) => <Icon label="AddShoppingCart" data-testid={id} />

const init = (props?: ButtonProps, role = 'button') => {
  renderWithTheme(<Button {...props}>Won Games</Button>)
  return screen.getByRole(role, { name: /won games/i })
}

describe('<Button />', () => {
  it('should render the correct snapshot', () => {
    const { container } = renderWithTheme(<Button>Won Games</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render medium size by default', () => {
    const sut = init()
    expect(sut).toHaveStyle({
      fontSize: theme.font.sizes.medium,
      padding: theme.spacing(1, 4)
    })
  })

  it('should render small size', () => {
    const sut = init({ size: 'small' })
    expect(sut).toHaveStyle({
      fontSize: theme.font.sizes.small,
      padding: theme.spacing(1)
    })
  })

  it('should render large size', () => {
    const sut = init({ size: 'large' })
    expect(sut).toHaveStyle({
      fontSize: theme.font.sizes.large,
      padding: theme.spacing(1, 6)
    })
  })

  it('should render fullWidth', () => {
    const sut = init({ fullWidth: true })
    expect(sut).toHaveStyle({
      width: '100%'
    })
  })

  it('should render an icon at start', () => {
    const sut = init({ startIcon: makeIcon('startIcon') })
    expect(sut).toBeInTheDocument()
    expect(screen.getByTestId('startIcon')).toBeInTheDocument()
  })

  it('should render an icon in the end', () => {
    const sut = init({ endIcon: makeIcon('endIcon') })
    expect(sut).toBeInTheDocument()
    expect(screen.getByTestId('endIcon')).toBeInTheDocument()
  })

  it('should render as a link', () => {
    const sut = init({ as: 'a', href: '/link' }, 'link')
    expect(sut).toHaveAttribute('href', '/link')
  })

  it('should render a text button', () => {
    const sut = init({ text: true })
    expect(sut).toHaveStyle({
      background: 'none',
      color: theme.colors.primary
    })
  })
})
