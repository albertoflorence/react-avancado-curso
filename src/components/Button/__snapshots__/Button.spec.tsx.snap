// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the correct snapshot 1`] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: none;
  border-radius: 4px;
  padding: 8px;
  cursor: pointer;
  font-size: 1rem;
  padding: 8px 32px;
}

.c0:hover {
  -webkit-filter: brightness(95%);
  filter: brightness(95%);
}

<button
  class="c0"
>
  <span>
     
    Won Games
     
  </span>
</button>
`;
